### PR TITLE
Simple and intuitive fullscreen behavior

### PR DESCRIPTION
**Pull request by RubenVerborgh**
The current Shower behavior is that it stores in local storage whether or not a slideshow was last displayed in full screen. This is counterintuitive and continues to annoy me on several devices; you just never get what you expect (but instead what you happened to have last time).

This commit fixes that by proposing the following simple and intuitive decision logic:
- http://example.org/slides/ takes the user to the list of slides.
- http://example.org/slides/#27 takes the user to slide 27 in full screen.

I think it's all we ever need; local storage introduced unpredictability and confusion.

To me, this proposal is the intuitive behavior, because it allows me to send links to others for either the presentation as a whole, or a specific slide. It also fixes the annoyance that you currently can't exit full screen mode on a keyboardless device.

### CODE REVIEW

**Comment by pepelsbey** [comment]
I’m afraid this is too radical. I really like the ability to go though slides in list mode and link to a certain slide in a context or next/previous slides in list mode.

There’s another idea how to get full screen mode state in URL back.

- http://example.com/#27
- http://example.com/#27:full

And in each case an actual ID exists so the slide still could get a `:target` state:

- `<section id="27">`
- `<section id="27:full">`

This `:full` thing should applended to each slide in full screen mode.

What do you think?

**Comment by pepelsbey** [comment]
And this is a valid ID value:

image

**Comment by RubenVerborgh** [comment]
> I really like the ability to go though slides in list mode and link to a certain slide in a context or next/previous slides in list mode.

I never found the necessity to link to slide in list mode, because it is way less obvious that you're linking to a specific slide. There is a highlight, but it is only minimal (and rightly so).

But I understand you want this functionality.

> This :full thing should applended to each slide in full screen mode.

> What do you think?

Following what I wrote above about obviousness, how about the opposite?

- `#27` for a full-screen slide
- `#list:27` or `#27:list` for a slide in list mode

It seems just more logical to me that, if you link to a slide, you see the _slide_ (only, full), not the list that contains that slide, with a minimal highlight on that slide.

**Comment by pepelsbey** [comment]
I think one of the main differences of Shower is that it focuses on slides list first. Most of the other slide decks start from the full screen by default and it’s usually not clear that there’s overview mode (if there’s one at all). You always have to go forward slide by slide though boring introduction looking for something interesting. In Shower case it’s also possible by `<body class="shower full">`, but **it’s not default**. Default is the list and I’d really like to keep it this way. Meaning that `:full` should be an extra parameter in URL, not `:list`.

**Comment by RubenVerborgh** [comment]
> Default is the list and I’d really like to keep it this way.

I understand, but if I want people to see the list, I'll link them to the list.
If I want them to see the slide, I'll link them to the slide.

I.e., I agree on list focus (and that's a great feature of Shower), but it perhaps shouldn't be overdone. If I like to the slide, I really mean slide, not list.

That said, I can proceed in any way you want—my main annoyance was local state.
I'm not sure about the necessity of changing ids in full-screen mode though. Is `target` that important, given that they would be JavaScript-activated anyway?

**Comment by pepelsbey** [comment]
I think I agree that changing ID on the fly is a bit overkill and such full screen routing is a must have feature. But if you remember there used to be a `?full` query parameter to keep full screen state in a link. It required History API that Chrome decided to disallow for `file:///` protocol. And presentations are often opened from flash drives without any local or remote servers. That was the reson to move to the `sessionStorage` full screen states. AFAIK, fragment identifiers in URL doesn’t require any `history.*` API calls and still could be changed for local files via JS. But I’d double check it before going any futher.

**Comment by pepelsbey** [comment]
FYI, I have merged new Selenium tests I’ve been working on recently to the `master`. There are 43 of them passing and 4 pending. It’s not tested in Travis yet as it requires Selenium grid and I’m still figuring out where to get it, preferably for free (Sauce Labs, etc.) But you can still test it locally, just `npm install` and `npm run test`.

Hopefully this will help developing core, I really appreciate your help.

**Comment by RubenVerborgh** [comment]
That's great news, thanks! I'll try to make an improved proposal for this PR soon.

**Comment by RubenVerborgh** [comment]
This new proposal has one link: `#37` point to slide 37, and opens by default in list view.

**Comment by RubenVerborgh** [comment]
Some context for this new proposal:
- The fact that `#37` points to list view respects the Shower defaults.
- We lose the ability to link specifically to a full or list-view mode of a slide. However, I think this ability is too complex for most users. People likely just want to link to a slide; they would not know the link is different whether they are in list or full mode, and cannot be expected to switch before they copy the link.

What do you think, @pepelsbey?

**Comment by pepelsbey** [comment]
But how’s that different from the current behaviour apart from `sessionStorage` that keeps full stare between reloads? Sometimes reload is very useful for resetting all `next` steps.

Sorry for not following up on this for a while, I’m having a busy time.

**Comment by RubenVerborgh** [comment]
> But how’s that different from the current behaviour apart from sessionStorage that keeps full state between reloads?

This PR indeed removes sessionStorage, so link always open consistently. (My main issue from a usability behavior was that links to slides were counterintuitive/unpredictable.)

**Comment by pepelsbey** [comment]
But what should I do if I want to just reload my presentation? Imagine it’s already open, you’re halfway through and you realize that something’s not right with your demo. I usually just hit Cmd R and instantly see updated slide in full mode. Your proposal would get me back to list mode which is rather bad.

**Comment by RubenVerborgh** [comment]
That's right; this is why my initial proposal was to always go to full-screen mode when a slide is linked directly.

Mmm, I don't know… seems like we do need different links to full-screen and list-view slides after all. _However_, I expect that the average user browsing slides will always do so in full-screen mode, and hence always share full-screen links. And sharing a list-view link is hard: you'd first have to click the slide (which opens full-screen mode), escape full-screen mode, and then copy.

Long story short: with different links for full-screen and list-view slides, users will end up sharing the full-screen link anyway. So maybe my proposal to always open individual slides in full-screen mode (but a link to the entire presentation in list mode) wasn't so bad after all.

LocalStorage is the current alternative, but I find its unpredictability rather annoying.

**Comment by pepelsbey** [comment]
> I expect that the average user browsing slides
> will always do so in full-screen mode

I doubt that. Because it’s like reading a book where only one sentence is written on a single page. It’s just too slow. Full mode is good for presenting because it allows you to guide audience through your story step by step, switching to the next one when you’re ready. But you don’t need it when you go through slides looking for a link or funny picture you saw yesterday.

Personally, I always read my slides in list mode because they are big enough to read even on a tiny 12” display. I go to full mode only for presentation. So this _full mode first_ would break this habit and I believe it’s not just me.

**Comment by RubenVerborgh** [comment]
But how would you (or anyone) share a link to a specific slide from list mode?

**Comment by pepelsbey** [comment]
Currently there’s a current slide indicator in both built-in themes: thick blue outline or darker shadow. It gives you an idea of what slide is current and you can go forward by pressing Space, Enter, arrows, etc and URL will match the current slide. I know, it’s not optimal but it works: you can navigate to a slide you want and get URL from address bar.

We could also wrap every slide to a link to itself: `<a href="#27"><section id="27"/></a>` just position it on top of it `<section id="27"/><a href="#27"></a>` to make it possible to copy a link to a slide. But I’m not sure it would be too useful.

**Comment by RubenVerborgh** [comment]
Different way of thinking: how about we make this configurable with:

- `linkMode: 'last'` (default, uses local storage to open link in last used mode)
- `linkMode: 'list'` (opens links in list view)
- `linkMode: 'full'` (opens links in full screen view)

**Comment by shvaikalesh** [comment]
Chrome 59, `history.{push,replace}State` work in `file:///` protocol. Should we go back to query param? Should we add an opt-out to disable `sessionStorage` behavior?

**Comment by pepelsbey** [comment]
Seriously? Did they revert this behaviour? I don’t know… I got used to this behaviour :) But it would be good to be able to have a link with mode built-in.

There’s an [issue for that](https://github.com/shower/shower/issues/222) in the main repository.

**Comment by pepelsbey** [comment]
There’s a plan to reimplement `?full` state in URL, since Chrome suddenly decided to support History API in local files. So you’ll again have a way to link to full screen state.

**Comment by RubenVerborgh** [comment]
Good news, thanks!
